bootstrap/src/sun/nio/ch/lincheck/EventTracker.java

```diff
@@ -32,6 +32,10 @@ public interface EventTracker {
     void beforeNewObjectCreation(String className);
     void afterNewObjectCreation(Object obj);
 
+    void updateSnapshotOnFieldAccess(Object obj, String className, String fieldName, int codeLocation);
+    void updateSnapshotOnArrayElementAccess(Object array, int index, int codeLocation);
+    void updateSnapshotWithEnergeticTracking(Object[] objs);
```

Comment:
Let's not expose `updateSnapshot` functions on `EventTracker` level, and instead call them from existing functions like `beforeReadField`, etc. 

The reasons for that are as follows. 

1) Snapshot collection and restoring logic is an implementation detail (of `ManagedStrategy`). Different strategies may either not need it at all, or implement their own different algorithms. 

2) The signature of `updateSnapshotOnFieldAccess` and `beforeReadField` is very similar. 

3) The code of `SharedMemoryAcessTransformer` and `SnapshotTrackerTransformer` that handles instructions like `GETFIELD`, `PUTFIELD`, etc, is very similar. In the long run, it would be more difficult to maintain two similar versions of APIs and transformer implementations.

Reply:
The `updateSnapshotWithEnergeticTracking` should be exposed, since I need to call it from the `Injections` class and there I cannot check for `getEventTracker() instanceof ManagedStrategy`, thus, I need to have this method in the `EventTracker` interface